test/WeakKeyIdDict-test.jl

```diff
@@ -0,0 +1,157 @@
+    @test WeakKeyIdDict(Pair(A, 2)) == wkd
+    @test !isempty(wkd)
+
+    wkd[A] = 42
+    @test wkd[A] == 42
+
+    wkd = WeakKeyIdDict(A=>2, B=>3, C=>4)
+    map!(x -> x + 1, values(wkd))
+    @test WeakKeyIdDict(A=>3, B=>4, C=>5) == wkd
+
+    wkd = WeakKeyIdDict(A=>2, B=>3, C=>4)
+    @test delete!(wkd, A) == WeakKeyIdDict(B=>3, C=>4)
+    @test delete!(wkd, A) == WeakKeyIdDict(B=>3, C=>4)  # deleting the same key twice works
+    @test delete!(wkd, C) == WeakKeyIdDict(B=>3)
+    @test delete!(wkd, B) == WeakKeyIdDict()
+    # adding stuff back is OK
+    wkd[A] = 2
+    wkd[B] = 3
+    wkd[C] = 4
+    @test wkd == WeakKeyIdDict(A=>2, B=>3, C=>4)
+
+    wkd = WeakKeyIdDict(A=>2)
+    @test get(wkd, A, 17) == 2
+    @test get!(wkd, A, 17) == 2
+    @test get(wkd, B, 17) == 17
+    @test length(wkd) == 1
+    @test get!(wkd, B, 17) == 17
+    @test length(wkd) == 2
+
+    wkd = WeakKeyIdDict(A=>2)
+    @test get(() -> 23, wkd, A) == 2
+    @test get!(() -> 23, wkd, A) == 2
+    @test get(() -> 23, wkd, B) == 23
+    @test length(wkd) == 1
+    @test get!(() -> 23, wkd, B) == 23
+    @test length(wkd) == 2
+
+    wkd = empty!(wkd)
+    @test wkd == empty(wkd)
+    @test wkd == empty(wkd)
+    @test typeof(wkd) == typeof(empty(wkd))
+    @test length(wkd) == 0
+    @test isempty(wkd)
+    @test isa(wkd, WeakKeyIdDict)
+    @test WeakKeyIdDict() == WeakKeyIdDict(())
+
+    @test_throws ArgumentError WeakKeyIdDict([1, 2, 3])
+
+    # WeakKeyIdDict does not convert keys
+    @test_throws ArgumentError WeakKeyIdDict{Int,Any}(5.0=>1)
+
+    # iterator
+    wkd = WeakKeyIdDict(A=>2, B=>3, C=>4)
+    @test Set(collect(wkd)) == Set([A=>2, B=>3, C=>4])
+    @test 2+3+4 == sum(v for (k,v) in wkd)
+
+    # WeakKeyIdDict hashes with object-id
+    AA = copy(A)
+    GC.@preserve A AA begin
+        wkd = WeakKeyIdDict(A=>1, AA=>2)
+        @test length(wkd)==2
+        kk = collect(keys(wkd))
+        @test kk[1]==kk[2]
+        @test kk[1]!==kk[2]
+    end
+
+    # WeakKeyIdDict compares to other dicts:
+    @test IdDict(A=>1)!=WeakKeyIdDict(A=>1)
+    @test Dict(A=>1)==WeakKeyIdDict(A=>1)
+    @test Dict(copy(A)=>1)!=WeakKeyIdDict(A=>1)
+
+    # issue #26939
+    d26939 = WeakKeyIdDict()
+    d26939[big"1.0" + 1.1] = 1
+    GC.gc() # make sure this doesn't segfault
+
+end
+
+# The following tests involve the garbage collector and
+# don't work if they are inside a @testset
+wkd = WeakKeyIdDict([42]=>2, [43]=>3, [44]=>4)
+for k in keys(wkd)
+  delete!(wkd, k)
+end
+@test isempty(wkd)
+GC.gc()  # try to get it to evict some weak references
+@test isempty(wkd)
+
+# verify that garbage collection takes care of our weak references
+A = [1]
+wkd = WeakKeyIdDict(A => 1)
+let tmp = [ 42 ]
+  @test length(wkd) == 1
+  wkd[tmp] = 2
+  @test length(wkd) == 2
+end
+# at this point there is no strong reference left to the vector [42]
+# previously reachable via tmp
+GC.gc(true)
+
+@test length(wkd) == 1
+@test length(keys(wkd)) == 1
+@test WeakKeyIdDict(A => 1) == wkd
```

Comment:
Assuming the CI tests pass now, I am fine with how high the coverage now is (> 90%). Note this ugliness: I could not find a way to get those tests which interact with the GC to work "right" while inside a `@testset`. But I am not aware of a serious drawback of the above, other than it not being reported in the output of `Pkg.test("AbstractAlgbera")` ... ?

Reply:
Does it work if you put it into a `let` scope? Currently, CI fails as there are two global objects both named `A` in the tests.